internal/api/user.go

```diff
@@ -122,37 +103,86 @@ func (a *API) UserUpdate(w http.ResponseWriter, r *http.Request) error {
-				if session == nil || now.After(session.CreatedAt.Add(24*time.Hour)) {
-					if len(params.Nonce) == 0 {
-						return badRequestError("Password update requires reauthentication")
-					}
-					if terr = a.verifyReauthentication(params.Nonce, tx, config, user); terr != nil {
-						return terr
-					}
+	if params.AppData != nil {
+		if !isAdmin(user, config) {
+			return unauthorizedError("Updating app_metadata requires admin privileges")
+		}
+	}
+
+	if user.IsSSOUser {
+		updatingForbiddenFields := false
+
+		updatingForbiddenFields = updatingForbiddenFields || params.Password != nil && *params.Password != ""
+		updatingForbiddenFields = updatingForbiddenFields || params.Email != "" && params.Email != user.GetEmail()
+		updatingForbiddenFields = updatingForbiddenFields || params.Phone != "" && params.Phone != user.GetPhone()
+		updatingForbiddenFields = updatingForbiddenFields || params.Nonce != ""
+
+		if updatingForbiddenFields {
+			return unprocessableEntityError("Updating email, phone, password of a SSO account only possible via SSO")
+		}
+	}
+
+	if params.Email != "" && user.GetEmail() != params.Email {
+		if duplicateUser, err := models.IsDuplicatedEmail(db, params.Email, aud, user); err != nil {
+			return internalServerError("Database error checking email").WithInternalError(err)
+		} else if duplicateUser != nil {
+			return unprocessableEntityError(DuplicateEmailMsg)
+		}
+	}
+
+	if params.Phone != "" && user.GetPhone() != params.Phone {
+		if exists, err := models.IsDuplicatedPhone(db, params.Phone, aud); err != nil {
+			return internalServerError("Database error checking phone").WithInternalError(err)
+		} else if exists {
+			return unprocessableEntityError(DuplicatePhoneMsg)
+		}
+	}
+
+	if params.Password != nil {
+		if config.Security.UpdatePasswordRequireReauthentication {
+			now := time.Now()
+			// we require reauthentication if the user hasn't signed in recently in the current session
+			if session == nil || now.After(session.CreatedAt.Add(24*time.Hour)) {
+				if len(params.Nonce) == 0 {
+					return badRequestError("Password update requires reauthentication")
 				}
+				if err := a.verifyReauthentication(params.Nonce, db, config, user); err != nil {
+					return err
+				}
+			}
+		}
+
+		password := *params.Password
+
+		// empty string means remove the password
+		if password != "" {
+			if user.EncryptedPassword != "" && user.Authenticate(ctx, password) {
+				return unprocessableEntityError("New password should be different from the old password.")
 			}
+		}
 
+		if err := user.SetPassword(ctx, password); err != nil {
+			return err
+		}
+	}
+
+	err = db.Transaction(func(tx *storage.Connection) error {
+		var terr error
+		if params.Password != nil {
 			var sessionID *uuid.UUID
 			if session != nil {
 				sessionID = &session.ID
 			}
 
-			if terr = user.UpdatePassword(tx, *params.Password, sessionID); terr != nil {
+			if terr = user.UpdatePassword(tx, sessionID); terr != nil {
```

Comment:
`user.UpdatePassword` no longer does the password hashing, you must call `user.SetPassword` first which modifies `EncryptedPassword` and then do the update in a transaction. This removes the problem where the database pool can be exhausted by sending too many user update requests with different passwords.